src/components/encryption.ts

```diff
@@ -0,0 +1,283 @@
+    }
+};
+
+interface DedupePresence {
+    userId: string;
+    currentlyActive?: boolean;
+    presence: "online"|"offline"|"unavailable";
+    statusMsg?: string;
+    ts: number;
+}
+/**
+ * The EncryptedEventBroker ensures that we provide a single encrypted
+ * event to bridges from potentially multiple /sync responses. The broker
+ * is also responsible for starting these syncs depending on which users
+ * can read the room.
+ *
+ * More broadly speaking, the bridge handles encrypted events currently by
+ * listening over the AS stream for encrypted messages, and then spinning
+ * up a /sync in order to read the message. In order to decrypt them, we
+ * proxy these requests through https://github.com/matrix-org/pantalaimon.
+ *
+ *   +-------------------+
+ *   |  Homeserver       |
+ *   +--------+----------+
+ *           ^
+ *           | Proxy
+ *           |
+ *           |
+ *   +--------+----------+
+ *   |  Pantalaimon      |
+ *   +--------+----------+
+ *           ^ /sync requests
+ *           |
+ *           |
+ *   +--------+----------+
+ *   |  Bridge           |
+ *   +-------------------+
+ *
+ * We also gain things like presence, read receipts and typing for free.
+ */
+export class EncryptedEventBroker {
+    constructor(
+        private membership: MembershipCache,
+        private asBot: AppServiceBot,
+        private onEvent: (weakEvent: WeakEvent) => void,
+        private onEphemeralEvent: (event: EphemeralEvent) => void,
+        private getIntent: (userId: string) => Intent
+        ) {
+
+        this.presenceCleanupInterval = setInterval(() => {
+            const ts = Date.now() - 30000;
+            this.receivedPresence = this.receivedPresence.filter(
+                presence => presence.ts < ts
+            );
+        }, 15000);
+    }
+    private receivedPresence: DedupePresence[] = [];
+    private presenceCleanupInterval: NodeJS.Timeout;
+    private handledEvents = new Set<string>();
+    private userForRoom = new Map<string, string>();
+
+    private eventsPendingSync = new Set<string>();
+    // We should probably make these LRUs eventually
+    private eventsPendingAS: WeakEvent[] = [];
+
+    private syncingClients = new Set<any>();
+
+    /**
+     * Called when the bridge gets an event through an appservice transaction.
+     * @param event
+     * @returns Should the event be passthrough
+     */
+    public async onASEvent(event: WeakEvent) {
+        if (event.type === "m.room.member" && event.state_key && event.content.membership) {
+            const existingSyncUser = this.userForRoom.get(event.room_id);
+            if (existingSyncUser === event.state_key && event.content.membership !== "join") {
+                // User has left the room (or are banned/invited), they are no longer our sync targets.
+                this.userForRoom.delete(event.room_id);
+                // TODO: Stop syncing entirely?
+            }
+            // TODO: If join, should we pre-emptively sync.
```

Comment:
Are you aware that you left some TODOs here?

Reply:
Having looked at this, neither seem trivial to implement and can probably wait for another time.